Comments to understand functions in the backend

cplkake opened the pull request:
Adds comments describing what the following functions are set up to accomplish above each respective function:

server/src/server.ts
- ifDefinedSet
- getUidForApiKey
- doApiKeyBasicAuth
- doApiKeyAuth
- doXidApiKeyAuth
- doHeaderAuth
- recordPermanentCookieZidJoin
- doXidConversationIdAuth
- _auth
  - getKey
  - doAuth
- enableAgid
- redirectIfHasZidButNoConversationId
- getXids
- handle_GET_xids
- doCookieAuth
- isOwner

server/src/session.ts
- getUserInfoForSessionToken

server/src/user.ts
- createDummyUser
- getXidRecordByXidOwnerId

colinmegill committed:
Streaming exports (#1826)

* Switch to non-native Postgres client.

And add a "streaming" API for making database queries, which streams the
results from the database to Node as they are generated by Postgres.

This allows Node to process the rows one by one (and garbage collect in
between), which is much easier on the VM when we need to do big queries that
summarize data (or just format it and incrementally spit it out an HTTP
response).

* Mostly refactoring.

This moves the handle_GET_reportExport route into its own file, which
necessitated refactoring some other things (zinvite and pca) out of server.ts
as well. Chipping away at the monolith.

This also converts the votes.csv report to use the streaming query from
Postgres, which is mostly a smoke test. It seems to work, so next I'll convert
it to stream the results incrementally to the HTTP response as well.

* Split each report into separate function.

* Count up comment votes in single pass over votes table.

There was actually a bug in the old SQL that aggregated votes from _all_
conversations instead of just the conversation in question, which is why it
took 30 seconds to run. With that bug fixed, even the super slow "do a full
subquery for each comment row" was actually quite fast. But this is way
cheaper/faster.

* Add participant-votes.csv export.

* Switch to non-native Postgres client.

And add a "streaming" API for making database queries, which streams the
results from the database to Node as they are generated by Postgres.

This allows Node to process the rows one by one (and garbage collect in
between), which is much easier on the VM when we need to do big queries that
summarize data (or just format it and incrementally spit it out an HTTP
response).

* Mostly refactoring.

This moves the handle_GET_reportExport route into its own file, which
necessitated refactoring some other things (zinvite and pca) out of server.ts
as well. Chipping away at the monolith.

This also converts the votes.csv report to use the streaming query from
Postgres, which is mostly a smoke test. It seems to work, so next I'll convert
it to stream the results incrementally to the HTTP response as well.

* Split each report into separate function.

* Count up comment votes in single pass over votes table.

There was actually a bug in the old SQL that aggregated votes from _all_
conversations instead of just the conversation in question, which is why it
took 30 seconds to run. With that bug fixed, even the super slow "do a full
subquery for each comment row" was actually quite fast. But this is way
cheaper/faster.

* Add participant-votes.csv export.

* Flip vote polarity.

In the raw votes table, -1 means agree and 1 means disagree, so we need to
count things correctly. And when exporting votes in participant votes, we flip
the sign so that 1 means agree and -1 means disagree.

* Properly escape comment text.

* add votes matrix, show data license preprod, logging.

---------

Co-authored-by: Michael Bayne <mdb@samskivert.com>

ballPointPenguin committed:
Revert "Streaming exports (#1826)"

This reverts commit 61d2940a5ff53a755240d661f94d430eecc8b543.

cplkake committed:
Comments to understand functions

cplkake committed:
Merge branch 'edge' of https://github.com/cplkake/polis into edge

NewJerseyStyle commented on server/src/server.ts:
```diff
@@ -234,6 +234,7 @@ function haltOnTimeout(req: { timedout: any }, res: any, next: () => void) {
   }
 }
 
+// checks if a property (name) exists in a source object and copies that property to a dest object
```
To enable IDEs recognize the comment and shows the comment when developer hover on the function where it is been used, the following syntax would be better:

```ts
/**
 * Checks if a property exists in a source object and copies it to a destination object.
 *
 * @param name - The name of the property to check and copy.
 * @param source - The source object containing the property.
 * @param dest - The destination object to copy the property to.  If undefined, a new object will be created.
 * @returns The destination object with the property copied if it exists in the source object. Returns undefined if the source object is undefined.
 * @throws {Error} If the source object is not an object or the destination object is not an object or undefined.
 **/
```

More about TSDoc specification. [https://tsdoc.org/](https://tsdoc.org/)

Zen-cronic replied:
Second this!

NewJerseyStyle commented on server/src/server.ts:
```diff
@@ -2,7 +2,7 @@
 "use strict";
 
-import akismetLib from "akismet";
+import akismetLib from "akismet";   // spam protection for user-submitted text
```
Adding a comment block at the beginning of your file to explain the purpose and key features of imported libraries is a good practice, especially when those libraries are not immediately obvious from the code itself or are less commonly used.

Here's how you might structure such a comment block at the beginning of the file, using the `akismet` example:
```ts
/**
 * @file This file contains the core logic for [brief description of the file's purpose].
 *
 * Dependencies:
 *  - `akismet`: This library is used for spam protection.  It provides functionality to check if user-submitted text is spam.  See [link to library documentation] for details on its API and usage.  Specifically, we're using it to [explain how you're using it, e.g., "check comments before posting"].
 *  - [Other library]: [brief explanation of its purpose and how it's used in this file].
 *
 *  Note: Ensure that the Akismet API key is correctly configured in the environment variables.  See [link to relevant documentation or configuration file].
 */

import { Client as AkismetClient } from "akismet";

// ... rest of your code ...
```

NewJerseyStyle commented:
It is trivial to manually edit all the comment for each function, you can used AI to assist you drafting the comment block and then review, refine it before commit.

patcon commented:
I'm going to close this for a sec, then re-open. that's just so I can try to sync the `edge` branch via the GitHub UI. Apparently creating PR's with commits from upstream confuses github and hides the "sync fork" option in the GitHub UI 🤷 

<img width="933" alt="Screenshot 2024-10-29 at 12 07 59 AM" src="https://github.com/user-attachments/assets/19583c7e-4f10-4a69-b54f-cc025fc19b0c">

patcon commented:
k that worked! I could click "sync fork" once I closed this. A bit wonky, but helpful to know :) Sorry for the noise in this PR

NewJerseyStyle commented:
I guess it is because Venish was using `edge` from `compdemocracy` to make this PR? It doesn't break anything except confusing GitHub UI? 😂

Zen-cronic commented:
After implementing the TSDoc, it's all good to merge!

NewJerseyStyle committed:
Update server.ts with block comments for functions

block comments generated by Gemini 1.5 Flash, detailed review is suggested (I roughly read through)

NewJerseyStyle committed:
Add block comment for getUserInfoForSessionToken

Comment generated by Gemini 1.5 Flash, more review is suggested

NewJerseyStyle committed:
Update user.ts adding block comments for functions

Comments generated by Gemini 1.5 Flash, more review suggested

NewJerseyStyle reviewed (approved):
I have updated comment to make it follow TSDoc should be good to merge now
